confmap/internal/mapstructure/encoder.go

```diff
@@ -0,0 +1,234 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mapstructure // import "go.opentelemetry.io/collector/confmap/internal/mapstructure"
```

Comment:
I was expected this to be a Decode into `map[string]interface{}`. Can you explain why the whole thing is needed?

Reply:
`mapstructure.Decode` into `map[string]interface{}` doesn't support the mapstructure tags of nested structs. It'll only marshal the top level string key and then since the output map value is an interface, it'll just set the value (https://github.com/mitchellh/mapstructure/blob/bf980b35cac4dfd34e05254ee5aba086504c3f96/mapstructure.go#L509). It decodes based on the structure of the `result` in the decoder config.

I think `mapstructure` in general isn't really designed for encoding.

Reply:
Not convinced based on https://github.com/mitchellh/mapstructure/issues/53. Are you sure, can you write this to that issue with a test to prove?

Reply:
I understand why that does not work. I think we can have a "DecodeHook" that if the "from" is struct then it generates from to be map[string]interface I think that may work.

What do you think?

Reply:
I think the main issue is more that the "to" reflect.Value is the bottleneck. It won't continue to run decode against the nested structs unless there's a corresponding "to" to decode into. If the "to" is just an interface, then it'll stop there. Unless we can modify the "to" in the "DecodeHook", I don't think a "DecodeHook" will help. I'll put an example in the issue you linked above when I get the chance.

Reply:
I was a bit mistaken. After a bit more testing it turns out that it's actually the `decodeMap` function that won't work well with the `TextMarshalerHookFunc`. Turns out it doesn't call `decode` on the map keys for nested fields if it is marshalling a struct into a map, so the hook doesn't get called (https://github.com/mitchellh/mapstructure/blob/bf980b35cac4dfd34e05254ee5aba086504c3f96/mapstructure.go#L1006). This is a problem for us because the `service.Config` is just a bunch of maps with non-string (`ComponentID`) keys that need to be marshalled.

https://go.dev/play/p/0YBJAW-iCtP

It also turns out that running `Decode` with a struct that has a marshaller that marshals it into a string will break the `mapstructure` decoder. It looks like the other ticket is closed, but I'll put the findings in it as well.

Reply:
I think this is more of the issue https://github.com/mitchellh/mapstructure/issues/166

Reply:
Doesn't look like the mapstructure repo is very active. Don't think I'll be getting a response soon. Would it be possible to proceed with this workaround and then if the mapstructure repo is able to address the issue, replace the implementation? The implementation isn't exposed to the user, so there shouldn't be any issues with future backwards compatibility.